### DISCUSSION TITLE
How can you use existing code in a notebook?

### DISCUSSION THREAD

**Discussion by jewpaltz** [dotnet/interactive#3046] [Q&A] [answered]
B"H

I would like to reference an existing .net project in my notebook.
Most of all, I'd like to call my existing EFCore model which has all of my mappings and business logic in it. But more broadly, I'd loke to call functions and instantiate classes that I have in my own projects.

I tried `#r ".\bin\Debug\net7.0\MyWebApp.dll"` Which has my EFCore model in it.
Then I tried instantiating the context with
```
using Microsoft.EntityFrameworkCore;

var options = new 
DbContextOptionsBuilder<MyNamespace.MyContext>()
                .UseSqlServer(CONNECTION_STRING, x=> x.UseNetTopologySuite()).Options;
var db = new MyNamespace.MyContext(options);
```
but I am getting quite a few errors.
Does anyone have any advice for me?

**Comment by jonsequitur** [comment, marked as answer]
It's hard to say without knowing which errors you're getting but I'm going to guess you're missing a bunch of assembly references. This approach can work but you'll also need `#r` statements for all of the transitive dependencies of MyWebApp.dll, which can be a little complicated. 

We do want to add direct support for this (#890) but it's been on the backlog for a while.

**Reply by jewpaltz** [reply]
B"H

Correct! Yes #890 is exactly what I was looking for. I hope that you are able to add that soon.
I see myself using both scenarios that you mention there. I would also add a sub-scenario to "defining methods to be shared" which is - using the notebook for exploratory data work as I'm developing an application.

Thank you for making a great project, and congrats for getting python back into the mix.

**Reply by jonsequitur** [reply]
Thanks! Glad it's proving useful.

> I would also add a sub-scenario to "defining methods to be shared" which is - using the notebook for exploratory data work as I'm developing an application.

If you'd like to add a comment elaborating on this idea in #890, that would be helpful.
